File: c/src/proto/lineage_replace_session_binding.c
#include "auth/lineage_replace_session_binding.h"

#include <stddef.h>
#include <string.h>

lineage_replace_session_binding_decision_t lineage_replace_classify_session_binding(
    const auth_v3_context_t *context,
    const lineage_replace_session_binding_expectation_t *expectation) {
    if (context == NULL || expectation == NULL || !expectation->auth_completion_verified)
        return LINEAGE_REPLACE_SESSION_REJECT_COMPLETION;
    if (context->protocol_version != expectation->expected_protocol_version)
        return LINEAGE_REPLACE_SESSION_REJECT_VERSION;
    if (context->suite_id != expectation->expected_suite_id)
        return LINEAGE_REPLACE_SESSION_REJECT_SUITE;
    if (context->profile_id != expectation->expected_profile_id)
        return LINEAGE_REPLACE_SESSION_REJECT_PROFILE;
    if (memcmp(context->session_id, expectation->expected_session_id, AUTH_SESSION_ID_LEN) != 0)
        return LINEAGE_REPLACE_SESSION_REJECT_SESSION_ID;
    if (memcmp(context->authz_context_hash, expectation->expected_authz_context_hash, AUTH_HASH_LEN) != 0)
        return LINEAGE_REPLACE_SESSION_REJECT_AUTHZ_CONTEXT;
    if (memcmp(context->channel_binding_hash, expectation->expected_channel_binding_hash, AUTH_HASH_LEN) != 0)
        return LINEAGE_REPLACE_SESSION_REJECT_CHANNEL_BINDING;
    return LINEAGE_REPLACE_SESSION_BOUND;
}

File: rust/crates/proto/src/lineage_replace_session_binding.rs
use crate::auth_v3::AuthV3Context;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LineageReplaceSessionBindingDecision {
    Bound,
    RejectCompletion,
    RejectVersion,
    RejectSuite,
    RejectProfile,
    RejectSessionId,
    RejectAuthzContext,
    RejectChannelBinding,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LineageReplaceSessionBindingExpectation {
    pub auth_completion_verified: bool,
    pub expected_protocol_version: u8,
    pub expected_suite_id: u16,
    pub expected_profile_id: u16,
    pub expected_session_id: [u8; 16],
    pub expected_authz_context_hash: [u8; 32],
    pub expected_channel_binding_hash: [u8; 32],
}

pub fn classify_lineage_replace_session_binding(
    context: Option<&AuthV3Context<'_>>,
    expectation: Option<&LineageReplaceSessionBindingExpectation>,
) -> LineageReplaceSessionBindingDecision {
    let (Some(context), Some(expectation)) = (context, expectation) else {
        return LineageReplaceSessionBindingDecision::RejectCompletion;
    };
    if !expectation.auth_completion_verified {
        return LineageReplaceSessionBindingDecision::RejectCompletion;
    }
    if context.protocol_version != expectation.expected_protocol_version {
        return LineageReplaceSessionBindingDecision::RejectVersion;
    }
    if context.suite_id != expectation.expected_suite_id {
        return LineageReplaceSessionBindingDecision::RejectSuite;
    }
    if context.profile_id != expectation.expected_profile_id {
        return LineageReplaceSessionBindingDecision::RejectProfile;
    }
    if *context.session_id != expectation.expected_session_id {
        return LineageReplaceSessionBindingDecision::RejectSessionId;
    }
    if *context.authz_context_hash != expectation.expected_authz_context_hash {
        return LineageReplaceSessionBindingDecision::RejectAuthzContext;
    }
    if *context.channel_binding_hash != expectation.expected_channel_binding_hash {
        return LineageReplaceSessionBindingDecision::RejectChannelBinding;
    }
    LineageReplaceSessionBindingDecision::Bound
}
